jac/jaclang/compiler/parser.py

```diff
@@ -915,112 +896,79 @@ def abstract_ability(self, kid: list[ast.AstNode]) -> ast.Ability:
+            is_static = self.match_token(Tok.KW_STATIC) is not None
+            self.consume_token(Tok.KW_CAN)
+            access = self.match(ast.SubTag)
+            semstr = self.match(ast.String)
+            name = self.consume(ast.NameAtom)
+            signature = self.match(ast.FuncSignature) or self.consume(
+                ast.EventSignature
+            )
+            return ast.Ability(
+                name_ref=name,
+                is_async=False,
+                is_override=is_override,
+                is_static=is_static,
+                is_abstract=True,
+                access=access,
+                semstr=semstr,
+                signature=signature,
+                body=None,
+                kid=kid,
             )
-            chomp = chomp[1:] if is_static else chomp
-            chomp = chomp[1:]
-            access = chomp[0] if isinstance(chomp[0], ast.SubTag) else None
-            chomp = chomp[1:] if access else chomp
-            semstr = chomp[0] if isinstance(chomp[0], ast.String) else None
-            chomp = chomp[1:] if semstr else chomp
-            name = chomp[0]
-            chomp = chomp[1:]
-            signature = chomp[0]
-            chomp = chomp[1:]
-            if isinstance(name, ast.NameAtom) and isinstance(
-                signature, (ast.FuncSignature, ast.EventSignature)
-            ):
-                return ast.Ability(
-                    name_ref=name,
-                    is_async=False,
-                    is_override=is_override,
-                    is_static=is_static,
-                    is_abstract=True,
-                    access=access,
-                    semstr=semstr,
-                    signature=signature,
-                    body=None,
-                    kid=kid,
-                )
-            else:
-                raise self.ice()
 
         def genai_ability(self, kid: list[ast.AstNode]) -> ast.Ability:
             """Grammar rule.
 
             genai_ability: KW_OVERRIDE? KW_STATIC? KW_CAN access_tag? STRING?
             named_ref (func_decl) KW_BY atomic_call SEMI
             """
-            chomp = [*kid]
-            is_override = (
-                isinstance(chomp[0], ast.Token) and chomp[0].name == Tok.KW_OVERRIDE
-            )
-            chomp = chomp[1:] if is_override else chomp
-            is_static = (
-                isinstance(chomp[0], ast.Token) and chomp[0].name == Tok.KW_STATIC
+            is_override = self.match_token(Tok.KW_OVERRIDE) is not None
+            is_static = self.match_token(Tok.KW_STATIC) is not None
+            self.consume_token(Tok.KW_CAN)
+            access = self.match(ast.SubTag)
+            semstr = self.match(ast.String)
+            name = self.consume(ast.NameAtom)
+            signature = self.match(ast.FuncSignature) or self.consume(
+                ast.EventSignature
+            )
+            self.consume_token(Tok.KW_BY)
+            body = self.consume(ast.FuncCall)
```

Comment:
Also consume token `SEMI`, the rule is make sure all the tokens/ rules are processed and thus at the end of this function (future) will be doing an assertion for that. LIKE:

```
parseing the rule.
assert self.nodes is empty. # <-- semicollon is not consumed an error.
```